### ISSUE TITLE
QUERY: Query body normalization is a bad idea

### ISSUE THREAD

**Issue by nicowilliams** [httpwg/http-extensions#2896]
See discussion in [HN thread](https://news.ycombinator.com/item?id=41552804).

**Comment by nicowilliams** [comment]
It might be a good idea to have a `NORMALIZE` method for requesting that the server normalize a `QUERY` body.

Basically, client-side normalization is going to be heavily dependent on the query MIME type.  Do MIME registrations even provide normalization functions?  Well, the registrations are fairly free-form, so that question is probably not relevant.

When JSON normalization has been discussed before it's proven to be an extremely tricky topic.

_Server-side_ normalization, however, is much more likely to be possible, especially if the server parses a query into an AST, as then a canonical representation of the AST should be possible without having to specify any details in any RFCs.

**Comment by reschke** [comment]
Are you suggesting to add text about normalization? Precisely where?

**Comment by reschke** [comment]
I disagree.

This is an implementation detail. A server can normalize based on how it uses the media type. No public spec is required here.

**Comment by nicowilliams** [comment]
Yes, the _server_ can normalize however it wants and it's a local detail for it.  But caches need not be located in the server, and _caches_ may not be able to normalize the same way as the _servers_.

**Comment by rafageist** [comment]
I’d like to expand on the semantic nature of QUERY. Given the name, it implies querying an endpoint, but how do we semantically define what constitutes a "query"? Without clear guidelines, there’s potential for abuse, as QUERY could be used for anything. Unlike GET, POST, PUT, or PATCH, which are generic, QUERY sounds specific, but the ambiguity opens the door to misuse. Additionally, with many possible query languages (e.g., SQL, GraphQL), if the protocol doesn’t define this, as it does with URIs (which have a well-defined query syntax), maybe QUERY isn’t the best name.

**Comment by reschke** [comment]
It *is* generic. It's safe method that can, contrary to GET, use a body.

Specific semantics will be defined based on media types of the request body.

**Comment by MikeBishop** [comment]
@rafageist, if you disagree with the name, please open a separate issue. That's not the topic of this issue.

**Comment by rafageist** [comment]
> @rafageist, if you disagree with the name, please open a separate issue. That's not the topic of this issue.

Ok, nut the title say "Query body normalization is a bad idea", and maybe "Fetch body normalization is a good idea"

**Comment by reschke** [comment]
The ticket is about what is currently in the spec, and that is QUERY. Can we please stop the nitpicking?

As Mike said, if you want to discuss the method name, open a separate ticket. Before you do so, please have a look at https://github.com/httpwg/http-extensions/issues/1614 before.

**Comment by reschke** [comment]
@nicowilliams , I'd like to get this discussion resolved.

First of all, once a server responds with a Location field, a cache can simply cache this like GET to that URI, and the issue goes away. (right?)

I also don't think it would be good to discourage caches from normalizing. The three types of normalization seem to be pretty safe to me. If a server treats those type of normalisations differently, that would essentially be a very stupid thing for the server to do.

**Comment by reschke** [comment]
Also, would it be correct to say that this is not about normalization in general, but specifically to normalisations outside the origin server?

**Comment by mnot** [comment]
@nicowilliams it would be helpful to understand the motivation for the change. Pointing to a huge HN thread requires the reader to do an unreasonable amount of work - can you summarize?

**Comment by mnot** [comment]
Discussed offline with @reschke. Absent more information, we think the best path forward is to add a security consideration along the lines of:

> Caches that normalize QUERY content incorrectly or in ways that are significantly different than how the server processes the content can return the incorrect response if normalization results in a false positive.

**Comment by nicowilliams** [comment]
> Caches that normalize QUERY content incorrectly or in ways that are significantly different than how the server processes the content can return the incorrect response if normalization results in a false positive.

If the cache is shared this can cause cache poisoning.

> Note that any such normalization is performed solely for the purpose of generating a cache key; it does not change the request itself.

If the cache normalizes incorrectly then this can yield aliasing issues.

What I don't understand is: how shall caches even normalize?

> Normalizing based upon knowledge of the semantics of the content itself, as indicated by the request's Content-Type field.

Hmm, so do MIME type registrations indicate how to normalize documents?  Must they now do so for content types that are suitable for use in `QUERY` request bodies?

**Comment by reschke** [comment]
>>    Caches that normalize QUERY content incorrectly or in ways that are significantly different than how the server processes the content can return the incorrect response if normalization results in a false positive.
>
> If the cache is shared this can cause cache poisoning.

So clarify that this would be specifically bad  for shared caches?

>> Note that any such normalization is performed solely for the purpose of generating a cache key; it does not change the request itself.
>
> If the cache normalizes incorrectly then this can yield aliasing issues.

Please explain "aliasing" in this context.

> >    Normalizing based upon knowledge of the semantics of the content itself, as indicated by the request's Content-Type field.
>
> Hmm, so do MIME type registrations indicate how to normalize documents? Must they now do so for content types that are suitable for use in QUERY request bodies?

Some. No.

**Comment by nicowilliams** [comment]
> > > Caches that normalize QUERY content incorrectly or in ways that are significantly different than how the server processes the content can return the incorrect response if normalization results in a false positive.
> > 
> > 
> > If the cache is shared this can cause cache poisoning.
> 
> So clarify that this would be specifically bad for shared caches?

If two different queries with different semantics get normalized to the same query then one user-agent can poison the cache for another.  That's what I meant by aliasing.  I suppose this seems unlikely.

Maybe I can rephrase my question entirely in terms of MIME types:

How shall a cache implementor normalize a `QUERY` request body?  What references shall they look at?  The MIME type registry's registration entries appear to be completely mum on this topic.  JSON will commonly be used I imagine, but neither JSON nor I-JSON specify a canonical or deterministic encoding option.  CBOR does specify a deterministic encoding option (which used to be referred to as canonical), so presumably MIME types that make use of CBOR can have `QUERY` bodies normalized by caches.

Perhaps JSON can be normalized to a degree: remove interstitial whitespace, canonicalize strings so that only characters that must be escaped are, and... do something about number representations.  Normalization is a private details of the cache implementation I suppose.  As long as the implementors do not incorrectly implement normalization I guess it's Ok.  Fine.

**Comment by vidraj** [comment]
(Note: I didn't read the HN thread, as the link doesn't work for me. Perhaps I'm just repeating ideas that were already discussed there. Also, apologies for resurrecting a closed issue – I hope it's better than creating a new one for what's essentially the same problem.)

I think the spec should contain some advisory text for how to opt out of such normalization. Basically any service that parses user requests and returns detailed error information in the (fairly common) form of “line X, column Y: Incorrect ABC, expected DEF” would be negatively impacted by it, because any normalization is all but guaranteed to change these positions, even if it doesn't change the semantics of the processing.

This goes double for services which are specifically supposed to process potentially incorrect data, such as validators, or services which depend on the usually insignificant details, such as code/data formatters, formatting checkers or linters.

What is the best way of opting out? I can see three options:

1. Use POST instead of QUERY, with all the downsides that QUERY is supposed to avoid.
2. Specify a different content type, such as `application/octet-stream` and pass the actual type in a different manner, e.g. an `X-Actual-Content-Type` header (ugly!).
3. Specify the `Cache-Control: no-transform` header in the request. I'm not sure whether this falls under the intended usage of the header, so if you believe it does, I believe the query-method spec should specifically point out that proxies MUST honor the header also for purposes of caching.

**Comment by vidraj** [comment]
I'll try to explain the issue with an example, hopefully that'll get my point across better than an abstract description. O:-)

---

Let's assume that two clients, `A` and `B`, are querying a server through a common cache. The server has an endpoint, `/sum`, which calculates a sum of a JSON array of numbers.

1. Client `A` sends a query
    ````json
    [ 1, 2, "3", 4 ]
    ````
    The cache passes it on as-is (no transformation takes place, therefore `Cache-Control: no-transform` is not needed).
2. The server replies with a 400-class response saying `Error: string found on line 1, number expected`, with `Cache-Control: max-age 3600`. This is simplistic, but reasonsble error handling.
    The cache stores the response under a key containing (among other details) the transformed body `[1,2,"3",4]` with all whitespace removed.
3. Client `B` sends a query
    ````json
    [
        1,
        2,
        "3",
        4
    ]
    ````
    The cache looks it up internally under key `[1,2,"3",4]`. Since there is a fresh cached response, it doesn't pass the request further. 
4. The cache replies with `Error: string found on line 1, number expected`, a wrong message for this request – the actual error is on line 4.

**Comment by reschke** [comment]
So you're saying the cache caches the 4xx?

**Comment by reschke** [comment]
https://greenbytes.de/tech/webdav/rfc9110.html#overview.of.status.codes

So no, a 400 response by default is not cacheable.

**Comment by vidraj** [comment]
It's not heuristic caching here, the response has an explicit Cache-Control header.

From [RFC 9111](https://www.rfc-editor.org/rfc/rfc9111.html#name-storing-responses-in-caches) (excerpt):

> A cache MUST NOT store a response to a request unless:
> - the request method is understood by the cache;
> - the response status code is final […];
> - […] the cache understands the response status code;
> - the no-store cache directive is not present in the response […];
> - if the cache is shared: the private response directive is […] not present […];
> - if the cache is shared: the Authorization header field is not present […];
> - the response contains at least one of the following: […] a max-age response directive […]

All of these are fulfilled, so the cache may store the response. This hypothetical cache does.

**Comment by vidraj** [comment]
And this is not just a hypothetical – real-world implementations behave like this, at least according to the excellent <https://cache-tests.fyi/> page (under “An optimal HTTP cache reuses a fresh 400 response with explict freshness”).

**Comment by reschke** [comment]
OK.

So why would the origin server *want* to make the response cacheable?

Also, what's the probability that two clients send the same problematic request *and* actually have some way to process the response? IMHO that requires a human to understand and act on it.

This really seems like an edge case, which, FWIW, would not be specific to QUERY.

**Comment by vidraj** [comment]
> So why would the origin server want to make the response cacheable?

I have three answers to that:
1. Why wouldn't it? If you send the same request, you get the same response, so cache it. Allowing caching should be the default, rather than something exceptional.
2. It allows *the user-agent* to cache the response. Caching in HTTP is uniform in the sense that you can't target specific caches (except for the private/public distinction, which is useless here), so if you want to alleviate the load on your servers by allowing browsers to just give the user back the previous response if they mash the submit button, or if they have the page open for months while closing and restarting the browser, you have to allow caching at all levels.
3. I honestly thought the whole point of the QUERY proposal is to make more responses cacheable – it's a niche method, and the primary (niche) use is to allow clients to make cacheable *structured*, *long* queries. Unstructured and short queries fit into the URLs of GET requests, uncacheable queries can be made using POSTs. Structured short queries can be made using GETs, but it requires app-specific handling (e.g. `/sum?q=%5B1%2C2%2C%223%22%2C4%5D&qtype=text%2Fjson`).


> Also, what's the probability that two clients send the same problematic request and actually have some way to process the response? IMHO that requires a human to understand and act on it.

Yes, a human makes the request and is then misled by the response. Is that surprising? Error messages are often meant to be shown to humans (perhaps through a syslog of some kind), even when the happy path (with successful response) leads to automatic processing.

As for the probability: I already gave you an example yesterday that doesn't depend on there being multiple clients at all. One client makes a request, realizes there's an error in it, submits a reformatted query to get a more informative message, gets the same message regardless.

If this client is a “hacker” (can open browser devtools, knows how to work with them), they might be able to force their user-agent to send the request with `Cache-Control: no-cache` to make the cache reload the response from the server. But you can't really bake this into the application outright, otherwise you lose all other desirable effects of caching, and you can't expect ordinary users to do this.


> This really seems like an edge case, which, FWIW, would not be specific to QUERY.

Can you give me an example of another HTTP method that allows caches to serve mismatched responses? Also, yes, error handling is often considered only as an afterthought, if at all. I believe we should be trying to do better than that.

**Comment by vidraj** [comment]
> Also, what's the probability that two clients send the same problematic request _and_ actually have some way to process the response?

Another point to this: Issues that only happen probabilistically when the stars align *just right* are the worst to debug. If I had a service like this and an issue would happen once a month where a client would be served a mismatched response from the cache, I would switch back to POSTs.

Presumably you don't want to create a brand new standard where the first reaction of a seasoned sysadmin to the suggestion of using it is “yeah, that thing is kinda flakey, just avoid it”. :-)

**Comment by vidraj** [comment]
> Why would an origin server want to make a _4xx_ response cacheable?

Please see the part of my answer that you didn't quote. Caching errors doesn't seem particularly outlandish to me. Some 4xx responses are even heuristically cacheable, so presumably it didn't seem outlandish to the other authors of the HTTP specs either.


> > If this client is a “hacker” (can open browser devtools, knows how to work with them), they might be able to force their user-agent to send the request with `Cache-Control: no-cache` to make the cache reload the response from the server. But you can't really bake this into the application outright, otherwise you lose all other desirable effects of caching, and you can't expect ordinary users to do this.
> 
> And how likely is it that a client used by a "non hacker" would be able to read & understand the payload of a 4xx response?

Uhh… I don't understand where you're going with this. 400-class errors are typically shown to end users, yeah. The payload I gave as an example above was actually a human-readable string. I thought it is obvious that it's meant to be shown to the end user somehow, maybe in the same place where a successful response would appear.

But this is just one example – perhaps the response might even be a full HTML page with the error shown in red. 404 errors are typically shown to the user in this manner and users understand them well. See e.g. <https://github.com/httpwg/nonexistent-repo> for an example of what that might look like. Perhaps it's instead a JSON payload describing the error, meant to be interpreted by a web application (a JavaScript client?) which draws the offending part of the query in bold, with the error message as an explanatory bubble next to it.

I believe this is all standard practice. When you have an HTTP-enabled app and the backend returns an error, it also supplies an explanation that the frontend somehow makes available to the user. Details vary, of course, but I don't think they're particularly relevant here.


> > > This really seems like an edge case, which, FWIW, would not be specific to QUERY.
> > 
> > Can you give me an example of another HTTP method that allows caches to serve mismatched responses? Also, yes, error handling is often considered only as an afterthought, if at all. I believe we should be trying to do better than that.
> 
> We need to balance complexity of the protocol with that. I firmly believe that this isn't something we need to spec.

So… it *is* specific to QUERY?

Also, the additional complexity of the protocol is really a single sentence. “If the request contains `Cache-Control: no-transform`, cached responses to requests with transformed (not byte-for-byte identical) bodies MUST NOT be used to satisfy this request.” (Obviously, some slight rewording might be necessary, I'm not a native speaker and also not a particularly good writer.)

**Comment by reschke** [comment]
So the client is able to set cache-control to "no-transform" but not to "no-cache"?

I'm sorry, but I don't think this discussion is leading anywhere.

**Comment by reschke** [comment]
> > So, just to clarify: you want to disallow (or change the default) for normalization of the payload as cache key because
> 
> I want a way to opt out.
> 
> > ```
> > 1. for a broken payload, an origin server might send diagnostics on a 400 (or 422) response that do not make sense when returned for a different payload but with the same normalization, _and_
> > 
> > 2. that origin server chooses to make that response cacheable using cache-control, _and_
> > 
> > 3. the user of the client is able to see and understand the response content, would be able to fix the QUERY request content, but is not able to add cache-control to prevent caching?
> > ```
> 
> For a given user-supplied payload (form contents or the like), the server returns either a 200 OK response, or a 400 Bad Request error response, depending on whether the user-supplied query is syntactically correct or not. Both are cacheable according to the same rules, because why wouldn't they be? (This seems to be a fundamental difference between our points of view. I'm not sure why you think 400-class responses should be uncacheable in general, many are by default.)

Because the HTTP says so.

> ...

**Comment by vidraj** [comment]
Link please? I provided a quote above directly from the RFC which shows this is false, and another link that shows existing implementations cache `400 Bad Request` when instructed to do so. Your own link says that multiple 4xx responses even get heuristically cached by default.

**Comment by vidraj** [comment]
You seem to be focused on heuristic caching. That is a workaround that shouldn't be used in new applications. It was meant to help modern clients deal with old, or perhaps broken, servers. I'm not sure why you give its rules the weight you do. But don't take it from me, read [what MDN has to say](https://developer.mozilla.org/en-US/docs/Web/HTTP/Guides/Caching#heuristic_caching):

> HTTP is designed to cache as much as possible, so even if no Cache-Control is given, responses will get stored and reused if certain conditions are met. This is called heuristic caching.
> […]
> Heuristic caching is a workaround that came before Cache-Control support became widely adopted, and basically all responses should explicitly specify a Cache-Control header.

**Comment by reschke** [comment]
https://www.rfc-editor.org/rfc/rfc9110.html#section-15.1-2

**Comment by vidraj** [comment]
> https://www.rfc-editor.org/rfc/rfc9110.html#section-15.1-2

See my previous response. This only concerns the defaults, and the defaults are a legacy thing meant as a sane and somewhat conservative default for HTTP 1.0 servers. It imposes no restrictions on modern applications, which should (according to modern best practices) always provide explicit `Cache-Control`.

**Comment by reschke** [comment]
>  which should (according to modern best practices) always provide explicit Cache-Control

Maybe I'm missing something here. Where do I find descriptions of these "modern best practices"?

(cc @mnot as our in-house caching expert)

**Comment by mnot** [comment]
I don't think the overhead of adding an opt-out mechanism from normalisation is worth it here, especially considering how narrow the use case is. Without normalisation you lose most of the benefits of caching anyway, unless you have a highly homogenous client population -- which isn't the web as deployed. So, you might as well make the response uncacheable or use POST.

Furthermore, developers can (and will) adapt to avoid such assumptions in their error messages.

**Comment by nicowilliams** [comment]
@mnot @reschke what @vidraj and I are saying is that normalization of queries should not be a `MUST`.  It probably shouldn't even be a `SHOULD`.  It can be a `MAY`.

How do you normalize -let's say- SQL??  Really, I've no idea how to normalize SQL.  There are tools to tell you if two radically different queries are equivalent, so the query normalization problem for SQL is non-trivial.  Moreover SQL query normalization might require knowledge of the schema and application.  For example, `INSERT INTO x (a) SELECT 1 WHERE NOT EXISTS (SELECT a FROM x WHERE a = 1;` is idempotent and might be "safe" in some schemas (where `x` is metadata) and not others (where `x` is data).

I have no idea how to normalize JSON either.  We had over 1,000 messages in the threads leading up to RFC 7159's publication, and some dealt with the topic of normalization.  No one knows how to normalize real number literals for example.

I mean, you're asking the _caches_ to normalize.  If the caches normalize two semantically different queries onto the same cache key by accident then you'll have a serious problem.  You added a Security Considerations note about this, but all it says is that if you get bit by this then haha too bad because normalization is a `MUST` so you lose.

If it was the servers doing the normalization that might make more sense.  Maybe they could return a URI, much like a `POST` returning a `201` does, for the cache key corresponding to the client's query?  Then `QUERY` would begin to look a lot more like `POST`, just with the safety and idempotence of `GET`.  What's nice about letting the _server_ do the normalization and the assignment of a cache key is that the caches can be real dumb -- caches generally are real dumb, right?  Are there caches today that know how to interpret URI local-parts and q-params where those represent queries?  Why shouldn't caches also normalize `GET` URI local-parts and q-params when they [somehow] know the application?

What happens if a cache understands some query content types and not others?  Must it respond with a `4xx` to the ones it doesn't understand, or can it pass them along w/o normalization?  This question for me really does it: `REQUIRED` query normalization in the _cache_ is wishful thinking that cannot and will not materialize.

Can we at least have some examples of what this normalization would look like?  Can we have some examples of why we would need it?  I get that two users might type the same query in different ways, but so what.

Do you have running code to demonstrate this requirement or is the requirement aspirational?  If you do not have running code, have you sought commentary from cache authors and maintainers?

**Comment by reschke** [comment]
> [@mnot](https://github.com/mnot) [@reschke](https://github.com/reschke) what [@vidraj](https://github.com/vidraj) and I are saying is that normalization of queries should not be a `MUST`. It probably shouldn't even be a `SHOULD`. It could be a `MAY`, maybe maybe -- after writing the following I think it has to be a `MUST NOT`.

Indeed. That's why it's *not* a MUST. 

> How do you normalize -let's say- SQL?? Really, I've no idea how to normalize SQL. There are tools to tell you if two radically different queries are equivalent, so the query normalization problem for SQL is non-trivial. Moreover SQL query normalization might require knowledge of the schema and application. For example, `INSERT INTO x (a) SELECT 1 WHERE NOT EXISTS (SELECT a FROM x WHERE a = 1);` is idempotent and might be "safe" in some schemas (where `x` is metadata) and not others (where `x` is data).

I don't know either. Well, except removing useless whitespace.

But anyway, there's no MUST here.

> ...
> I mean, you're asking the _caches_ to normalize. If the caches normalize two semantically different queries onto the same cache key by accident then you'll have a serious problem. You added a Security Considerations note about this, but all it says is that if you get bit by this then haha too bad because normalization is a `MUST` so you lose.
> ...

It's not a "MUST".

> ...I
> What happens if a cache understands some query content types and not others? Must it respond with a `4xx` to the ones it doesn't understand, or can it pass them along w/o normalization? This question for me really does it: `REQUIRED` query normalization in the _cache_ is wishful thinking that cannot and will not materialize.
> ...

It is not "REQUIRED". Yes, it can pass it content along. There is no requirement to normalize.

> Can we at least have some examples of what this normalization would look like? Can we have some examples of why we would need it? I get that two users might type the same query in different ways, but so what.
> ...

Like:

- Removing content encoding(s) ([Section 8.4](https://greenbytes.de/tech/webdav/rfc9110.html#rfc.section.8.4) of [[HTTP]](https://greenbytes.de/tech/webdav/draft-ietf-httpbis-safe-method-w-body-latest.html#HTTP)).
-  Normalizing based upon knowledge of format conventions, as indicated by any media subtype suffix in the request's Content-Type field (e.g., "+json", see [Section 4.2.8](https://www.rfc-editor.org/rfc/rfc6838.html#section-4.2.8) of [[RFC6838]](https://greenbytes.de/tech/webdav/draft-ietf-httpbis-safe-method-w-body-latest.html#RFC6838)).
-  Normalizing based upon knowledge of the semantics of the content itself, as indicated by the request's Content-Type field.

???

> Do you have running code to demonstrate this requirement or is the requirement aspirational? If you do not have running code, have you sought commentary from cache authors and maintainers?

I believe this whole comment is based on an incorrect reading of the spec.

**Comment by reschke** [comment]
I do agree that `SHOULD` is a bit strange, if, what follows is a list of essentially optional things. I think that indeed a `MAY` would be correct (as it may not be obvious that a cache can do).

Let's open a separate ticket for this.

**Comment by nicowilliams** [comment]
Making it a `MAY` would go a great deal of the distance to assuaging my concerns :)

I'd still like to first note that there is a big difference between middleware applying codec conversions (e.g., content encoding compression function changes[^1]), or data serialization rules changes[^2]) and making changes to the _actual contents_ rather than its encodings.  The latter seems so unlikely to happen that I think you should also just consider removing all text about query normalization.

Back to the SQL normalization example, suppose the content-type were a JSON-encoded AST of the actual SQL.  Then we could say that removing interstitial whitespace would be a form of normalization, though what to do about normalization of real numbers would remain a thorny issue.  That would be very reasonable, but it would require a client application that could generate that AST.  And that brings me to: I think the best place to do normalization is on the user-agent!  If the user-agent is real dumb and the query is just taken from a text input widget on the page and user input, then the user-agent can't normalize, and really neither can any middleware -- only the server can, and the middleware can learn nothing about how to normalize future queries.  If the application is smart on the client side (e.g., it runs JavaScript programs that parse user-input queries and turn them into ASTs) then the application can also be smart enough to perform some sorts of normalization -- at least those sorts that are feasible.

Thus as I think about it I believe the only places where query normalization can be performed are:

- the client-side (and then the middleware need not know anything about query forms)
- the server-side, though this does not help middleware unless we had a verb for normalizing queries, then middleware could pre-flight a `NORMALIZEQUERY` request for every `QUERY` so that it could check the cache.

I think perhaps if query normalization is a valuable and feasible thing, and caching canonical queries were equally desirable, then you might want a `NORMALIZEQUERY` verb!

Even if query normalization were non-trivial, having a `NORMALIZEQUERY` verb might still be desirable, even if most applications would only ever produce a `501`.

[^1]: Trivial and application-independent.  Just decompress and re-compress.
[^2]: Less trivial because the middleware must be able to identify the application's request body content schema, unless the codec change were for a self-describing encoding like BER->DER, or DER->CER, etc.  But still feasible because the middleware need not understand the application's semantics, and the middleware need not be a part of the application.

**Comment by vidraj** [comment]
> I don't think the overhead of adding an opt-out mechanism from normalisation is worth it here, especially considering how narrow the use case is.

The opt-out mechanism already exists for transforming proxies and implementing it is a MUST (See [RFC 9110](https://www.rfc-editor.org/rfc/rfc9110#section-7.7)). For a good reason.

In my opinion, implementing this mechanism for QUERY is also already a MUST, you just didn't spell it out clearly, because any proxy that implements the cache key transformations as outlined in the QUERY spec is technically a transforming proxy from the point of view of the client. It's just a really bad one, because it sometimes performs transformations, sometimes it doesn't, and it transforms the content in a really weird way that sometimes optimizes it and sometimes pessimizes. And all of this happens without involving the server, so you might have a service which sometimes fails to work correctly, but the operators of the service don't see anything wrong on their end.

Perhaps another thing you should consider is to require that a proxy that transforms request bodies for the purpose of caching also sends the canonical representation to the server instead of the original. That way, the transformation is always applied, the server operators see its results and all clients get the same response no matter who sent their request first. This doesn't impose any additional requirements on the proxy beyond handling the first request in the same manner as subsequent ones (e.g. fully buffering the body before sending it on and calculating the transformed body immediately as opposed to lazily).

---

To summarize my argument why an opt-out is necessary in a general way: The semantics of a request body can't be inferred from the content type. A few examples:
- Whitespace in a JSON request is sometimes significant (because error messages want to point to an exact line+column).
- The exact bytes of any request are sometimes significant (because the service in question is a cryptographic digest calculator, to be used e.g. before contacting a banking API, which requires each request to contain a correct digest; or because the service *is* the banking API and requires correct digests to be present).
- Sometimes, the following `text/html` bodies are identical:
  ```html
  <p>Hello, world!
  ```
  and
  ```html
  <!DOCTYPE html>
  <html>
  <head><title></title></head>
  <body><p>Hello, world!</p></body>
  </html>
  ```
  But if the service in question accepts an HTML snippet and returns a full page with the snippet embedded, then embedding the second one makes the generated web page invalid. If the service checks that, it might send an uncacheable error back, but not all services do. Or perhaps the service was originally made to process `application/xhtml+xml`, but as an extension (and in accordance with Postel's law), it also allows clients to make HTML requests, as long as the text is trivially convertible to XHTML, which the latter is, but the former isn't.
- Sometimes, sending a JPEG image file with metadata, or sending the same pixels in a different JPEG file without metadata makes no difference. If the service is supposed to add a fancy border and burn any ImageTitle and Photographer tags into the bottom part, the metadata will make a difference. Both requests with metadata and without are fine, neither will result in an error, just an image without the desired title, or (perhaps worse) with a different title than intended.
- Sometimes, losslessly reencoding a JPEG image into JPEG-XL image is inconsequential. Not if you want to show the result in a browser, though, because browsers generally support JPEG, but don't support JPEG-XL. The service might process a JPEG-XL just fine, sending another JPEG-XL back, because it doesn't know whether you need it to be browser-compatible or not.
- Sometimes, a `text/plain` document with DOS-style CRLF newlines is equivalent to a document with UNIX-style LF newlines. But not if the service is a typesetter that requires the CRs to be present, otherwise it makes all text gradually shift right. This probably won't result in any errors either, just in a broken document.
- Sometimes, the C program
  ```C
  int main(void) {
      int x = 1;
      return 0;
  }
  ```
  is equivalent to the program
  ```C
  int main(void) {
      return 0;
  }
  ```
  and to
  ```C
  int main(void) {
      return 0;
      int x = 1;
  }
  ```
  But not when compiled with `-Wunused-variable -Werror`, because that makes the first and third program not compile, and not when the compiler processes C89 only, because there the third program is invalid. This might again not result in any errors getting returned from the back end, because if the service in question is Compiler Explorer, it's not there to *actually compile* your programs, it's there to *show you the compiler output*. Which it did, 200 OK. (Of course, it's difficult to imagine a web proxy that understands `text/x-csrc` at the required level, but it's allowed to.)

I could go on, but hopefully you get the gist of my argument.


> Without normalisation you lose most of the benefits of caching anyway, unless you have a highly homogenous client population -- which isn't the web as deployed. So, you might as well make the response uncacheable or use POST.

Counterpoints:
- One user typically only has a single client, so even in the worst case, caching in the user-agent will always work. The same argument can be applied to most forward proxies.
- Even big deployments with heterogeneous clients (YouTube? Facebook?), typically have one major official client (the JavaScript on the official web pages), potentially two (a desktop/mobile app), but both controlled by the same entity. They can just generate good requests in the first place. The small amount of not-very-well-cacheable requests coming from other clients can mostly be ignored.
- An open API (ActivityPub?) can have documented best practices that lead to optimal caching.

Also, this problem already exists with GET requests and the ordering of the query parameters and the developers who care solve it through documenting the preferred order. There were proposals for opt-in automatic reordering, but AFAIK they went nowhere, because making such transformations work well is difficult.


> Furthermore, developers can (and will) adapt to avoid such assumptions in their error messages.

I'd say that's impossible. Both technically impossible (in well-formed XML or JSON, you can identify error positions by their path, as e.g. XPath or `jq` would, but if the file is not well-formed, you already have a problem – and many file formats are not strictly tree-based and therefore don't have such a path mechanism) and socially impossible (good luck persuading the GCC authors to rewrite their compiler to use something else than file:line:column).


> I'd still like to first note that there is a big difference between middleware applying codec conversions (e.g., content encoding compression function changes1), or data serialization rules changes2) and making changes to the actual contents rather than its encodings. The latter seems so unlikely to happen that I think you should also just consider removing all text about query normalization.

I'd say even the former is problematic, because the proxy can't really know which encodings are supported by the server – this is one area where requests are very different from responses.

**Comment by vidraj** [comment]
> > > I'd still like to first note that there is a big difference between middleware applying codec conversions (e.g., content encoding compression function changes1), or data serialization rules changes2) and making changes to the actual contents rather than its encodings. The latter seems so unlikely to happen that I think you should also just consider removing all text about query normalization.
> 
> > I'd say even the former is problematic, because the proxy can't really know which encodings are supported by the server – this is one area where requests are very different from responses.
> 
> Sure it can:
> 
>     * user-agent sends a `GET` of some resource with `Accept-Encoding: br`
> 
>     * the proxy does not respond immediately but does a `GET` of the same to the origin with `Accept-Encoding: br, gzip` (etc.)
> 
>     * the server responds with `200` w/ `gzip` encoding
> 
>     * the proxy decompresses and then recompresses to `br` and responds with `200` and the re-compressed response body.
> 
> 
> (The server can also respond that there are no acceptable encodings in `br, gzip`.)

Yeah, with responses, this design space is explored well. It's not really a permitted transformation, but many proxies do it anyway and it mostly works. And since `TE: gzip` etc. never got off the ground, repurposing Content-Encoding for that purpose is a pragmatic decision.

But for request bodies, there is no `Accept-Encoding`. When I write a client that sends a request with a Content-Encoded body, I have to use some out-of-band protocol to learn which encodings are accepted by the server. Typically, that protocol is “reading the documentation” and the choice gets hardcoded into the software. This is viable for service-specific clients or perhaps reverse proxies, but not for general-purpose clients (the most common general-purpose clients – web browsers – AFAIK don't support compressed requests even with JavaScript scripting) and certainly not for forward proxies outside of some very specific deployments.

It would be theoretically possible to either:
1. use `Expect: 100-Continue` in the request to allow the server to send a `415 Unsupported Media Type` before sending the body, or
2. use an `OPTIONS` request to learn the possible representations ahead-of-time,

but both of these have limited support, especially in web browsers. I'm using `100 Continue` for server-to-server API access and it works well, the server-side support is there, but AFAIK browsers don't have a way of requesting it. I don't expect proxies to start doing either any time soon.


> > The opt-out mechanism already exists for transforming proxies and implementing it is a MUST (See [RFC 9110](https://www.rfc-editor.org/rfc/rfc9110#section-7.7)). For a good reason.
> 
> Sadly that is a _response_ transformation opt-out mechanism. I don't see an opt-out mechanism for _request_ bodies[.]

I believe that's an unintended omission or perhaps an outright typo, as the older [RFC 2616](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.9.5) specifies it for all messages.

(Tangent, not relevant for QUERY:)
That said, it's not clear from either RFC whether specifying `no-transform` in a request applies the directive to the response as well. RFC 9111 just says “avoid transforming the content”, but it's not specified *which* content it refers to. Since the exact same wording is used for both requests and responses (where it is clear), I assume this is, again, an omission and we should not read too much into it. There is probably no strong intent behind the wording there.


> > A proxy MUST NOT transform the content ([Section 6.4](https://www.rfc-editor.org/rfc/rfc9110#content)) of a response message that contains a no-transform cache directive ([Section 5.2.2.6](https://www.rfc-editor.org/rfc/rfc9111#section-5.2.2.6) of [[CACHING](https://www.rfc-editor.org/rfc/rfc9110#CACHING)]). Note that this does not apply to message transformations that do not change the content, such as the addition or removal of transfer codings ([Section 7](https://www.rfc-editor.org/rfc/rfc9112#section-7) of [[HTTP/1.1](https://www.rfc-editor.org/rfc/rfc9110#HTTP11)]).
> > A proxy MAY transform the content of a message that does not contain a no-transform cache directive. A proxy that transforms the content of a [200 (OK)](https://www.rfc-editor.org/rfc/rfc9110#status.200) response can inform downstream recipients that a transformation has been applied by changing the response status code to [203 (Non-Authoritative Information)](https://www.rfc-editor.org/rfc/rfc9110#status.203) ([Section 15.3.4](https://www.rfc-editor.org/rfc/rfc9110#status.203)).
> 
> I do think a _request_ content transformation opt-out mechanism is required, but it's very difficult to add such a thing now, and certainly it's difficult to do so _here_. (EDIT: I mean that adding a generic request body transformation opt-out to HTTP in an RFC that adds a verb might be _politically_ difficult.)

The mechanism is already present, proxies are already required to honor it (even if many don't – Opera Turbo certainly didn't, back in the day, and it was a problem), it's just necessary to make it clear in the QUERY spec that what the proxy is doing there is really a transformation (this is my interpretation, but it is what it looks like from the point of view of the client) with all it entails, including the requirement to honor these directives.

**Comment by nicowilliams** [comment]
> I believe that's an unintended omission or perhaps an outright typo, as the older [RFC 2616](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.9.5) specifies it for all messages.

Oh, sorry, I should have checked.  You're absolutely right then I think.  @reschke @mnot please see this.  There already is an opt-out.

**Comment by nicowilliams** [comment]
> But for request bodies, there is no `Accept-Encoding`.

Yes, for sure.  If you can't reconstruct the request body later then only `Expect`/`100-Continue` or `OPTIONS` can work, and those two are just not that universally supported.  This is a big problem in HTTP -- really, this and `401`/`3xx`-for-auth are the big problems in HTTP, IMO.
